app/services/events/stores/postgres/weighted_sum_query.rb

```diff
@@ -94,6 +111,91 @@ def period_ratio_sql
             END
           SQL
         end
+
+        def grouped_events_cte_sql(initial_values)
+          groups = store.grouped_by.map.with_index do |group, index|
+            "#{sanitized_propery_name(group)} AS g_#{index}"
+          end
+
+          <<-SQL
+            WITH events_data AS (
+              (#{grouped_initial_value_sql(initial_values)})
+              UNION
+              (#{
+                events
+                  .select("#{groups.join(', ')}, timestamp, (#{sanitized_propery_name})::numeric AS difference, events.created_at")
+                  .to_sql
+              })
+              UNION
+              (#{grouped_end_of_period_value_sql(initial_values)})
+            )
+          SQL
+        end
+
+        def grouped_initial_value_sql(initial_values)
+          values = initial_values.map do |initial_value|
+            groups = store.grouped_by.map do |g|
+              if initial_value[:groups][g]
+                "'#{ActiveRecord::Base.sanitize_sql_for_conditions(initial_value[:groups][g])}'"
+              else
+                'NULL'
+              end
+            end
+
+            "(#{groups.join(', ')}, timestamp without time zone :from_datetime, #{initial_value[:value]}, timestamp without time zone :from_datetime)"
```

Comment:
Layout/LineLength: Line is too long. [150/120] (https://rubystyle.guide#max-line-length)